### PR TITLE
[video_player] Increase seeking performance

### PR DESCRIPTION
**Pull request by henri2h**
This PR prevents calling the `seekTo` functions when the player is already buffering due to a previous `seekTo` call, this should enhance the seeking experience by not blocking the preview.

Fix: https://github.com/flutter/flutter/issues/101409

### What does it do ?

To prevent the issue described in the related PR, we want to limit the number of buffer not leading to an image being displayed.
In order to do this, we just ignore `seekTo` calls when the player is not ready to play.

**What if we ask for the final location when the preview is buffering ?**

This is the reason we store the asked position when we ignore the `seekTo` call, thus,
when the play is ready, i.e. it has displayed the new frame, we check if the user have asked for a final location
and render it if needed.

This allow to keep the player responsive even if multiple locations have been asked really quickly.

### Try proposed fix

 * clone https://github.com/fluttercommunity/chewie/pull/623
 * try to go through the video quickly using the progress bar

### Result

| With optimisation | Without |
| ------ | ------ |
| screen-20220331-182056 | screen-20220331-181554 |
|schnell|langsam|

*List which issues are fixed by this PR. You must list at least one issue.*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by Henri2h** [comment]
For the tests, not sure how to test this :(

**Comment by stuartmorgan** [comment]
> This PR prevents calling the `seekTo` functions when the player is already buffering due to a previous `seekTo` call, this should enhance the seeking experience by not blocking the preview.

Could you elaborate on how this is desirable? If someone seeks to position A, then immediately to position B, why would waiting until buffering is complete at position A to update the seek position when anything buffered at A isn't going to be used?

I recommend filing an issue about the underlying issue, since more discussion is needed and it's not obvious that this is the right solution.

**Comment by Henri2h** [comment]
> > This PR prevents calling the `seekTo` functions when the player is already buffering due to a previous `seekTo` call, this should enhance the seeking experience by not blocking the preview.
> 
> Could you elaborate on how this is desirable? If someone seeks to position A, then immediately to position B, why would waiting until buffering is complete at position A to update the seek position when anything buffered at A isn't going to be used?
> 
> I recommend filing an issue about the underlying issue, since more discussion is needed and it's not obvious that this is the right solution.

- [x] create issue
- [x] better explain the proposed solution

File: packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java [hunk 1]
```diff
@@ -255,8 +261,19 @@ void setPlaybackSpeed(double value) {
     exoPlayer.setPlaybackParameters(playbackParameters);
   }
 
+  int futureLoation = -1;
```
Comment (maheshmnj) [inline packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java:264]: This variable seems to have a typo can you correct it to `futureLocation`?

**Comment by stuartmorgan** [comment]
I'm confused about the relationship between the issue description and the PR. In the issue you describe the problem being about when the video *renders*, but here you're waiting on the *buffer* status. Those aren't the same thing. If the goal is to wait for a single frame to render before seeking again, why are you waiting for *buffering* to finish? Unless I'm missing something this will make seeking incredibly sluggish on slow network connections, and do a lot of pointless buffering (per my earlier comment).

**Comment by Henri2h** [comment]
Yes, you're right, I confused rendering and buffering in my explanation of the issue.

However, from the player desn't wait for the buffer to be full in order to continue playing (and in this case, unloking the `seekTo` function). The player ready event is triggerd only when the player is ready to play, event is buffering has not ended.
[playback-state-changes](https://exoplayer.dev/listening-to-player-events.html#playback-state-changes)

I'm listening for this event, as I know, it's the only way to know that the rendering did finish, i.e. something has been displayed on the screen.

Maybe, we could add a timeout function to unlock the `seekTo` function if the player didn't managed to buffer enough to be ready to play due to bad network connection ?

Concerning the pointless buffering, depend on what the user is doing
- when the user seek location A, then B immediately, then, yes, the player is going to wait for the player to be ready before seeking to the next position. However the timeout may help fix this issue to prevent the amount which will be buffered.
- when the user is scrolling through the video, the player is just constantly trying to buffer and discarding immediately. In this case, there is little difference between this PR and the acutal solution as in both cases, the player is constantly buffering. However,  this PR will help to not block totaly the display totaly and let the user know if he has to seek before or after when searching for a specific position.

**Comment by stuartmorgan** [comment]
> I'm listening for this event, because, as far as I know, it's the only way to know that something has been paint on the screen.

Couldn't you interpose the `Surface` to monitor frame available callbacks?

> * when the user seek location A, then B immediately, then, yes, the player is going to wait for the player to be ready before seeking to the next position. However, the timeout may help fix this issue to prevent the amount which will be buffered.

It won't fix it, it will just cap it by the timeout. If it takes, on average, X ms to be able to start playing from a new position (or time out), then without your PR it will start playing from B in X ms, but with your PR it will take 2X ms instead.

> * when the user is scrolling through the video, the player is just constantly trying to buffer and discarding immediately. In this case, there is little difference between this PR and the actual solution concerning pointeless buffering as in both cases

It is still going to be the case that it will increase the amount of time it takes to play back from the position where the seek bar is finally released by 50% on average.

> However,  this PR will help to not block the display totally and let the user know if he has to seek before or after when searching for a specific position by letting the player paint new content when seeking.

Couldn't you get the same effect without the delay at the end by rate limiting seek frequency on the Dart side instead? I.e., instead of sending continual seek events while dragging, send them only:
- at a throttled rate, and
- on drag completion

(It wouldn't actually wait for paint events, but if you add a timeout that wouldn't always be the case for the native solution either.)

**Comment by henri2h** [comment]
> Couldn't you get the same effect without the delay at the end by rate limiting seek frequency on the Dart side instead? I.e., instead of sending continual seek events while dragging, send them only:

Well, you're totally right, it's a more elegant solution. Just implemented it in https://github.com/fluttercommunity/chewie/pull/623 it's working great !
Thanks for your time.
I will close this PR has it's fix has been moved to the chewie plugin.

**Pull request closed without merging**